nodejs/eks/nodegroup.ts

```diff
@@ -677,8 +701,48 @@ export type ManagedNodeGroupOptions = Omit<aws.eks.NodeGroupArgs, "clusterName"
     scalingConfig?: pulumi.Input<awsInputs.eks.NodeGroupScalingConfig>
 };
 
-export function createManagedNodeGroup(name: string, args: ManagedNodeGroupOptions, parent: pulumi.ComponentResource): aws.eks.NodeGroup {
+export function createManagedNodeGroup(name: string, args: ManagedNodeGroupOptions): aws.eks.NodeGroup {
     const core = isCoreData(args.cluster) ? args.cluster : args.cluster.core;
+    const parent = isCoreData(args.cluster) ? args.cluster.cluster : args.cluster;
+
+    // Compute the nodegroup role.
+    let roleArn: pulumi.Output<string>;
+    if (!args.nodeRole && !args.nodeRoleArn) {
+        throw new Error(`An IAM role, or role ARN must be provided to create a managed node group`);
+    } else if (args.nodeRole && args.nodeRoleArn) {
+        throw new Error("nodeRole and nodeRoleArn are mutually exclusive to create a managed node group.");
+    } else if (args.nodeRole !== undefined) {
+        roleArn = pulumi.output(args.nodeRole).apply(r => r.arn);
+    } else if (args.nodeRoleArn !== undefined) {
+        roleArn = pulumi.output(args.nodeRoleArn);
+    } else {
+        roleArn = core.instanceRoles[0]?.arn;
```

Comment:
Isn't this case unreachable? It looks like you already covered all four possible cases.

Reply:
Done. This logic has been refactored.